File: LeetCodeSolutions/Python/0120-Triangle/solution.py
# https://leetcode.com/submissions/detail/721625743/
# Date of Submission: 2022-06-13

# Runtime: 87 ms, faster than 62.00% of Python3 online submissions for Triangle.
# Memory Usage: 14.7 MB, less than 99.00% of Python3 online submissions for Triangle.


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:

        currRow = 1

        # base cases
        if (len(triangle) == 1):
            return triangle[0][0]
        elif (len(triangle) == 2):
            return triangle[0][0] + min(triangle[1][0], triangle[1][1])

        # loop through and greedy lowest sum each column
        while (len(triangle) > 1):
            for i in range(1, currRow):
                triangle[1][i] += min(triangle[0][i - 1], triangle[0][i])
            triangle[1][0] += triangle[0][0]  # left edge case
            triangle[1][currRow] += triangle[0][currRow - 1]  # right edge case
            # reduce triangle
            currRow += 1
            triangle = triangle[1:]
        return min(triangle[0])  # triangle is now only 1 row tall, take min of that row


def getLeftEdgeSum(triangle) -> int:
    sum = 0
    for i in range(0, len(triangle)):
        sum += triangle[i][0]
    return sum


def getRightEdgeSum(triangle) -> int:
    sum = 0
    for i in range(0, len(triangle)):
        sum += triangle[i][i]
    return sum

File: LeetCodeSolutions/C/0120-Triangle/solution.c
/*

https://leetcode.com/submissions/detail/721639305/
Date of Submission: 2022-06-13

Runtime: 9 ms, faster than 65.83% of C online submissions for Triangle.
Memory Usage: 6.8 MB, less than 65.00% of C online submissions for Triangle.

*/

int min(int left, int right);

int minimumTotal(int** triangle, int triangleSize, int* triangleColSize){
    int currRow = 0;
    int lowestSum = 0;
    if (triangleSize == 1){
        return triangle[0][0];
    }
    else if (triangleSize == 2){
        return triangle[0][0] + min(triangle[1][0], triangle[1][1]);
    }

    while(currRow < triangleSize-1){
        currRow++;
        for(int i=1; i<currRow; i++){
            triangle[currRow][i] += min(triangle[currRow-1][i-1], triangle[currRow-1][i]);
        }
        triangle[currRow][0] += triangle[currRow-1][0];  // left edge case
        triangle[currRow][currRow] += triangle[currRow-1][currRow-1];  // right edge case
    }

    lowestSum = triangle[currRow][0];
    for (int i=1; i<currRow+1; i++){
        lowestSum = min(lowestSum,triangle[currRow][i]);
    }
    return lowestSum;
}

int min(int left, int right){
    if (left < right){
        return left;
    }

    return right;
}